sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala

```diff
@@ -237,17 +237,20 @@ class HiveExternalCatalogVersionsSuite extends SparkSubmitTestUtils {
 }
 
 object PROCESS_TABLES extends QueryTest with SQLTestUtils {
-  val releaseMirror = "https://dist.apache.org/repos/dist/release"
+  val releaseMirror = sys.env.getOrElse("SPARK_RELEASE_MIRROR",
+    "https://dist.apache.org/repos/dist/release")
   // Tests the latest version of every release line.
   val testingVersions: Seq[String] = {
     import scala.io.Source
     val versions: Seq[String] = try {
-      Source.fromURL(s"${releaseMirror}/spark").mkString
-        .split("\n")
-        .filter(_.contains("""<li><a href="spark-"""))
-        .filterNot(_.contains("preview"))
-        .map("""<a href="spark-(\d.\d.\d)/">""".r.findFirstMatchIn(_).get.group(1))
-        .filter(_ < org.apache.spark.SPARK_VERSION)
+      Utils.tryWithResource(Source.fromURL(s"$releaseMirror/spark")) { htmlContent =>
```

Comment:
`Source.fromURL(s"${releaseMirror}/spark").mkString` not close connection from `ource.fromURL(s"${releaseMirror}/spark")`.

Reply:
There are 29 `Source is not closed.` warn in Spark Code, I will create a jira to track this.

Reply:
Ya. Please revert from this PR and proceed in a separate PR. It's worth of it. Thanks.

Reply:
done~ I will fix all `Source is not closed` issue in a separate PR :)